### ISSUE TITLE
Library needs love

### ISSUE THREAD

**Issue by siscia** [charlesthk/python-mailchimp#56]
I don't want to sound mean or demanding or anything, just sharing some constructive feedback.

Unfortunately I wasn't able to use this library at all, and it is a shame :(

I am not even able to create a campaign and I don't know what I am doing wrong.

```
In [27]: client.campaign.create(data={"recipients":{"list_id":"3c307a9f3f"},"type":"regular","settings":{"subject_line":"Your Purchase Receipt","reply
    ...: _to":"orders@mammothhouse.com","from_name":"Customer Service"}})
---------------------------------------------------------------------------
HTTPError                                 Traceback (most recent call last)
<ipython-input-27-88505da45a16> in <module>()
----> 1 client.campaign.create(data={"recipients":{"list_id":"3c307a9f3f"},"type":"regular","settings":{"subject_line":"Your Purchase Receipt","reply_to":"orders@mammothhouse.com","from_name":"Customer Service"}})

/usr/local/lib/python2.7/dist-packages/mailchimp3/entities/campaign.pyc in create(self, data)
     22 
     23     def create(self, data):
---> 24         return self._mc_client._post(url=self.endpoint, data=data)
     25 
     26     def patch(self, campaign_id, data):

/usr/local/lib/python2.7/dist-packages/mailchimp3/mailchimpclient.pyc in wrapper(self, *args, **kwargs)
     23     def wrapper(self, *args, **kwargs):
     24         if self.enabled:
---> 25             return fn(self, *args, **kwargs)
     26     return wrapper
     27 

/usr/local/lib/python2.7/dist-packages/mailchimp3/mailchimpclient.pyc in _post(self, url, data)
     57             return "{'status': 204}"
     58         else:
---> 59             raise HTTPError(request=r)
     60 
     61     @_enabled_or_noop

HTTPError: 
```

What am I doing wrong here? Why it returns an error? How do I check what error I made?

Also, the version is not updated on pip.

It seems a reasonable wrapper but it is not usable without some form of feedback.

Please, take at least the time to update the version on pip or if you don't have time or interested (which is completely understandable), try to leave somewhere a looking for someone else to take charge.

Thanks anyway to your contribution to the open source.

Best,

S

**Comment by charlesthk** [comment]
Hi @siscia the repo has been updated this night. I put the new version on pypi today! 

Let me some time to check your error and I will get back to you. 

Charles

**Comment by stephenross** [comment]
Looking at the error message that you are receiving, it looks like you are using an older version of the python-mailchimp wrapper. I just merged an overhaul last night that restructured the wrapper to match the [MailChimp API documentation](http://developer.mailchimp.com/documentation/mailchimp/reference/overview/). Please download the newest version from here while we work on getting v2.0.0 up into PyPi and let us know if you still receive the error when you try to submit that request again. As a note, you will need to use `client.campaigns.create()` instead of the old `client.campaign.create()` when you are working with the 2.0.0 version of the wrapper. If you have any further issues, please let us know so that we can look into them.

**Comment by charlesthk** [comment]
The 2.0.0 version is now on Pypi.

**Comment by charlesthk** [comment]
@siscia is your problem resolved now ?

**Comment by siscia** [comment]
Huuummm, still something I am not getting.

```py
try:          
    client.campaigns.create(data={"recipients":[{"list_id":"ce0efa9e9d"}],"type":"regular","settings":{"subject_line":"Your Purchase Receipt","reply_to":"orders@mammothhouse.com","from_name":"Customer Service"}})
except Exception, e:
     pass

> e.request.body
 '{"type": "regular", "recipients": [{"list_id": "ce0efa9e9d"}], "settings": {"reply_to": "orders@mammothhouse.com", "subject_line": "Your Purchase Receipt", "from_name": "Customer Service"}}'

> e.response.json() 
{u'detail': u"The resource submitted could not be validated. For field-specific details, see the 'errors' array.",
 u'errors': [{u'field': u'recipients',
   u'message': u'Schema describes object, array found instead'}],
 u'instance': u'',
 u'status': 400,
 u'title': u'Invalid Resource',
 u'type': u'http://developer.mailchimp.com/documentation/mailchimp/guides/error-glossary/'}
```
Which is correct since the mailchimp api require the field `recipients` to be an object and not an array, however if I try to use an array I get:

```py
In [62]: client.campaigns.create(data={"recipients":{"list_id":"ce0efa9e9d"},"type":"regular","settings":{"subject_line":"Your Purchase Receipt","repl
    ...: y_to":"orders@mammothhouse.com","from_name":"Customer Service"}})
    ...: 
TypeError                                 Traceback (most recent call last)
<ipython-input-62-8ffe4884ab65> in <module>()
>> 1 client.campaigns.create(data={"recipients":{"list_id":"ce0efa9e9d"},"type":"regular","settings":{"subject_line":"Your Purchase Receipt","reply_to":"orders@mammothhouse.com","from_name":"Customer Service"}})

/usr/local/lib/python2.7/dist-packages/mailchimp3/entities/campaigns.pyc in create(self, data)
     74         for recipient in data['recipients']:
     75             try:
>> 76                 test = recipient['list_id']
     77             except KeyError as error:
     78                 error.message += ' The campaign recipient must have a list_id'

TypeError: string indices must be integers, not unicode

In [63]: client.campaigns.create(data={"recipients":{"list_id":["ce0efa9e9d"]},"type":"regular","settings":{"subject_line":"Your Purchase Receipt","re
    ...: ply_to":"orders@mammothhouse.com","from_name":"Customer Service"}})
    ...: 

TypeError                                 Traceback (most recent call last)
<ipython-input-63-14474671ae48> in <module>()
>> 1 client.campaigns.create(data={"recipients":{"list_id":["ce0efa9e9d"]},"type":"regular","settings":{"subject_line":"Your Purchase Receipt","reply_to":"orders@mammothhouse.com","from_name":"Customer Service"}})

/usr/local/lib/python2.7/dist-packages/mailchimp3/entities/campaigns.pyc in create(self, data)
     74         for recipient in data['recipients']:
     75             try:
>> 76                 test = recipient['list_id']
     77             except KeyError as error:
     78                 error.message += ' The campaign recipient must have a list_id'

TypeError: string indices must be integers, not unicode
```

I don't quite understand what is wrong...

**Comment by stephenross** [comment]
Hello @siscia, It looks like I messed up parsing that part of the campaign data. Please download the issue-56 branch and verify that that corrects the issue.

**Comment by stephenross** [comment]
I rebased to the current master and did a test run using the use case presented and was able to successfully create a campaign, so I have merged PR #64 and consider this issue resolved
